### PR TITLE
Bluetooth: Host: Add NULL check for addr in unpair

### PR DESCRIPTION
**Pull request by Thalley**
unpair may be called from bt_unpair where addr is NULL. One such case is the `bt clear all` shell command which calls bt_unpair with addr = NULL, and the addr is just forwarded to unpair which does not check for NULL.

bt_unpair allows for the addr to be NULL to clear all, but only if SMP is enabled.
    
Modified the checks in bt_unpair to increase readability and ensure that unpair is not called with NULL.

### CODE REVIEW

**Review by jori-nordic — DISMISSED** [review]

**Review by alwa-nordic — DISMISSED** [review]

**Review by jhedberg — CHANGES_REQUESTED** [review]
The idea with `bt_unpair(id, NULL)` is to never call `unpair()` directly, rather only through this foreach-loop:
```c
	if (IS_ENABLED(CONFIG_BT_SMP) &&
	    (!addr || bt_addr_le_eq(addr, BT_ADDR_LE_ANY))) {
		bt_foreach_bond(id, unpair_remote, &id);
		return 0;
	}
```

The only case where you'd end up going past this block and reaching the direct `unpair(id, NULL)` call is if `CONFIG_BT_SMP` is not enabled. I wonder what's the purpose of that, or if it's a bug and the function should not do anything at all if that Kconfig option is not set?

**Comment by jhedberg** [comment]
>  I wonder what's the purpose of that

Answering myself: unpairing is also supported for BR/EDR, however I don't think we support disabling LE and only enabling BR/EDR, so this is dead code. Anyway, I think the correct fix is to change the Kconfig conditions in such a way that not even static analysers think that the `unpair(id, NULL)` would be possible.

**Comment by Thalley** [comment]
> > I wonder what's the purpose of that
> 
> Answering myself: unpairing is also supported for BR/EDR, however I don't think we support disabling LE and only enabling BR/EDR, so this is dead code. Anyway, I think the correct fix is to change the Kconfig conditions in such a way that not even static analysers think that the `unpair(id, NULL)` would be possible.

You bring up some good points. I've refactored `bt_unpair` to only allow `addr == NULL` if `CONFIG_BT_SMP`. Not sure if any static analyzers have issues with my new suggestion, but it should work. 

In any case, this fix was actually an attempt to fix a runtime issue I found when using the shell, but I see now that the issue I saw will not be fixed by this (had a segmentation fault when doing `bt clear all`). Will investigate further.

**Review by jhedberg — APPROVED** [review]
This looks better, thanks. The only additional thing I might do is to clarify in `bt_foreach_bond()` documentation that it's only intended for LE bonds (e.g. it has no support for traversing available BR/EDR Link Keys), however that can be done in a future PR.

File: subsys/bluetooth/host/hci_core.c [hunk 1]
```diff
@@ -1949,13 +1950,21 @@ int bt_unpair(uint8_t id, const bt_addr_le_t *addr)
...
+		CHECKIF(addr == NULL) {
+			LOG_DBG("addr is NULL");
+			return -EINVAL;
+		}
+
+		unpair(id, addr);
 	}
 
-	unpair(id, addr);
 	return 0;
```
Comment (MariuszSkamra) [inline subsys/bluetooth/host/hci_core.c:1968]: Looks complicated, isn't it that `CHECKIF` is missing only?

```
	if (IS_ENABLED(CONFIG_BT_SMP) &&
	    (!addr || bt_addr_le_eq(addr, BT_ADDR_LE_ANY))) {
		bt_foreach_bond(id, unpair_remote, &id);
		return 0;
	}

        CHECKIF(addr == NULL) {
	        LOG_DBG("addr is NULL");
	        return -EINVAL;
        }

	unpair(id, addr);
	return 0;
```

Comment (MariuszSkamra) [reply subsys/bluetooth/host/hci_core.c:1968]: It's a nitpick :)
The linker is smart enough to remove unused code, please check how it looks for you, but for me the output assembly is exactly the same (`CONFIG_BT_SMP=y`, `CONFIG_ASSERT=y`).

**Review by MariuszSkamra — APPROVED** [review]

**Review by jori-nordic — APPROVED** [review]

**Pull request merged**